lib/redis_client/cluster/node/base_topology.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+class RedisClient
+  class Cluster
+    class Node
+      class BaseTopology
+        IGNORE_GENERIC_CONFIG_KEYS = %i[url host port path].freeze
+        attr_reader :clients, :primary_clients, :replica_clients
+
+        def initialize(pool, concurrent_worker, **kwargs)
+          @pool = pool
+          @clients = {}
+          @client_options = kwargs.reject { |k, _| IGNORE_GENERIC_CONFIG_KEYS.include?(k) }
+          @concurrent_worker = concurrent_worker
+          @replications = {}
+          @primary_node_keys = []
+          @replica_node_keys = []
+          @primary_clients = []
+          @replica_clients = []
```

Comment:
I think we can delay these initializations until they are needed. It seems to me that the allocations for the empty objects are superfluous.

Reply:
I'm happy to change this if you want, but initializing all the `@ivars` inside `#initialize` is a habit I'm getting into because it's something YJIT likes. This makes sure all instances of the class have the same object shape (because they all have the same set of ivars, regardless of what methods are called on them), and YJIT I think does (or will) treat new object shapes at a callsite as a trigger to deoptimise. In fact, Ruby will even print a warning about it if a class has too many different shape variations now: https://github.com/ruby/ruby/pull/9275.

This thing is only initialized once per cluster anyway now so I don't think the empty array allocation is a big deal.

Reply:
That makes sense for dynamic typed languages.

> I don't think the empty array allocation is a big deal.

Can we use the constant variables for lessening GC load? Is it meaningless?

```ruby
EMPTY_ARRAY = [].freeze
EMPTY_HASH = {}.freeze
```

Reply:
Ah yes I've seen this pattern elsewhere in the repo, I'll follow it.